Toolkit/ArcGISToolkit/TemplatePickerViewController.swift

```diff
@@ -0,0 +1,296 @@
+                                return
+                            }
+                            
+                            // update info with swatch
+                            infos[index].swatch = image
+                            
+                            // reload index where that info currently is
+                            if let index = strongSelf.indexPathForInfo(info){
+                                strongSelf.tableView.reloadRows(at: [index], with: .automatic)
+                            }
+                        }
+                    }
+                }
+            }
+            
+        }
+        
+    }
+    
+    // MARK: TableView delegate/datasource methods
+    
+    public func numberOfSectionsInTableView(_ tableView: UITableView) -> Int{
+        return tables.count
+    }
+    
+    public func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
+        return tables.count == 0 ? nil : tables[section].tableName
+    }
+    
+    public func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        
+        // when the user taps on a feature type
+        
+        // first get the selected object
+        let selectedFeatureTemplateInfo = infoForIndexPath(indexPath)!
+        
+        // If the search controller is still active, the delegate will not be
+        // able to dismiss us, if desired.
+        // Note: we can't do this before the above, or else the object we pull
+        // out of the datasource will be incorrect
+        navigationItem.searchController?.isActive = false
+        
+        delegate?.templatePickerViewControllerDidSelectTemplate(self, featureTemplateInfo: selectedFeatureTemplateInfo)
+    }
+    
+    override public func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        
+        guard tables.count > 0 else {
+            return 0
+        }
+        
+        let tableName = tables[section].tableName
+        let infos = self.currentDatasource[tableName]
+        return infos?.count ?? 0
+    }
+    
+    override public func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: cellReuseIdentifier, for: indexPath)
+        let info = infoForIndexPath(indexPath)!
+        cell.textLabel?.text = info.featureTemplate.name
+        cell.imageView?.image = info.swatch
+        return cell
+    }
+    
+    // MARK: go back, cancel methods
+    
+    @objc private func cancelAction(){
+        // If the search controller is still active, the delegate will not be
+        // able to dismiss us, if desired.
+        navigationItem.searchController?.isActive = false
+        delegate?.templatePickerViewControllerDidCancel(self)
+    }
+    
+    // MARK: IndexPath -> Info
+    
+    private func infoForIndexPath(_ indexPath: IndexPath) -> FeatureTemplateInfo?{
```

Comment:
I would suggest renaming this to the following:

```swift
func info(for indexPath: IndexPath) -> FeatureTemplateInfo
```

You're already force unwrapping the return value at every call site, so I would just do that in this function and return a non-optional. It wouldn't make sense for that there not be an info for a given index path.

Reply:
yeah - Mark said that as well - good call